File: mobile-android/app/src/main/java/com/example/p2paudio/protocol/VerificationCode.kt
package com.example.p2paudio.protocol

import java.security.MessageDigest

object VerificationCode {

    fun fromSessionAndFingerprints(
        sessionId: String,
        senderFingerprint: String,
        receiverFingerprint: String
    ): String {
        val normalized = listOf(sessionId, senderFingerprint, receiverFingerprint)
            .joinToString("|")
            .toByteArray(Charsets.UTF_8)
        val digest = MessageDigest.getInstance("SHA-256").digest(normalized)
        val numeric = ((digest[0].toLong() and 0xFFL) shl 24) or
            ((digest[1].toLong() and 0xFFL) shl 16) or
            ((digest[2].toLong() and 0xFFL) shl 8) or
            (digest[3].toLong() and 0xFFL)
        val value = numeric % 1_000_000L
        return value.toString().padStart(6, '0')
    }
}

File: desktop-windows/src/P2PAudio.Windows.Core/Protocol/VerificationCode.cs
using System.Security.Cryptography;
using System.Text;

namespace P2PAudio.Windows.Core.Protocol;

public static class VerificationCode
{
    public static string FromSessionAndFingerprints(
        string sessionId,
        string senderFingerprint,
        string receiverFingerprint
    )
    {
        var source = $"{sessionId}|{senderFingerprint}|{receiverFingerprint}";
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes(source));
        var numeric =
            ((uint)digest[0] << 24) |
            ((uint)digest[1] << 16) |
            ((uint)digest[2] << 8) |
            digest[3];
        var value = numeric % 1_000_000U;
        return value.ToString("D6");
    }
}
